pkg/status/handlers.go

```diff
@@ -63,9 +70,61 @@ func (a *API) version(w http.ResponseWriter, r *http.Request) {
 
 }
 
-func NewAPI(tracer trace.Tracer, monitor monitoring.MonitorInterface, logger logging.LoggerInterface) *API {
+func (a *API) deepCheck(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+
+	kready := make(chan bool, 1)
+	hready := make(chan bool, 1)
+
+	go func() {
```

Comment:
I went with the second option. Alessandro wants me to use waitgroups for blocking, and that makes it complicated to use the ctx.Done channel.

Reply:
the server will terminate if request takes too long

at the same time passing the context in can be done via

```

type Health struct {
   Hydra bool `json:"hydra"`
   Kratos bool `json:"kratos"`
}



func (a *API) deepCheck(w http.ResponseWriter, r *http.Request) {
    w.Header().Set("Content-Type", "application/json")
    w.WriteHeader(http.StatusOK)

    wg := sync.WaitGroup{}
    wg.Add(2)
    
    health := new(Health)


    go func(ctx context.Context) {
        status, err := a.service.CheckKratosReady(r.Context())
        if err != nil {
	    a.logger.Errorf("error when checking kratos status: %s", err)
            // false is the default of health.Kratos
            wg.Done()
            return  
        }
        
        health.Kratos = status
        
        wg.Done()
        
    }(r.Context())
    go func(ctx context.Context) {
        status, err := a.service.CheckHydraReady(r.Context())
        if err != nil {
	    a.logger.Errorf("error when checking hydra status: %s", err)
            // false is the default of health.Hydra
            wg.Done()
            return  
        }
        
        health.Hydra = status
        
        wg.Done()
        
    }(r.Context())
    
    wg.Wait()

    json.NewEncoder(w).Encode(health)
```

Reply:
two questions:
- is this implementing timeouts?
- is it functionally different for us to have a go routine defined inside the func, and reference a variable in the outer func (i.e. r.Context() ), compared to have the goroutine have a func parameter to which we pass the context object to?
- I doubt we're making a copy of the context object by passing them into a func as a parameter because Context is an interface.
Bonus:
- This code never uses the ctx function argument, it references the r.Context from the scope of the outer function.